rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,335 @@
+   that the job is owned exclusively by the worker. `WorkerLockId` will be
+   passed as an argument to other API functions below and will be used to
+   verify that the current worker is still the only worker executing that job.
+   `MaxPriority` is an optional value which can limit the maximum job priority
+   that will be accepted. Jobs with priorities higher than that will not be
+   accepted. The intended usage is to allow `Priority` to be used as a way to
+   schedule job for executing at a future date. For example, `Priority` MAY
+   indicate that a replication job which has been repeatedly failing should
+   not execute any sooner than one hour from now.
+
+```
+finish(Tx, Type, JobId, JobOpts, WorkerLockId) -> ok | worker_conflict | canceled
+```
+
+ - Called by the worker when the job has finished running. The `"data"` field
+   in `JobOpts` MAY contain a final result field or information about a fatal
+   error.
+
+```
+resubmit(Tx, Type, JobId, WorkerLockId) -> ok | worker_conflict | canceled
+
+```
+ - The worker MAY call this function in order to mark the job for
+   re-submission. This MAY be used to penalize jobs which have been running for
+   too long, or jobs which have been repeatedly failing. Note that both the
+   user and a worker can request a job to be re-submitted.
+
+```
+update(Tx, Type, JobId, JobOpts, WorkerLockId) -> ok | worker_conflict | canceled
+```
+ - This MAY be called to update a job's progress (update sequence, number of
+   changes left, etc.) This function MUST be called at least as often as the
+   `ActivityTimeout` in order for the activity monitor to not re-submit the
+   job into the pending queue due to inactivity.
+
+When the functions above return a `worker_conflict` value it means the activity
+monitor has already re-enqueued the job. It is now either pending or is executed
+by another worker. In that case, the caller MUST stop running the job, and they
+MUST NOT update the job's entry any longer.
+
+If `canceled` is returned, it means the user has requested the job to stop
+executing. In that case the worker MUST top running the job.
+
+#### Mutual Exclusion
+
+If each worker updates their job status on time, and the activity monitor is
+running correctly, the same job SHOULD be executed by no more than one worker
+at a time. But, if the worker process is blocked for too long (for instance, in
+an overload scenario), it may fail to update its status often enough, and the
+activity monitor MAY re-enqueue the job. However, even in such cases, the
+mutual exclusion constraint can be maintained as long as `update(Tx,...)`
+function is called in the same transaction as the type-specific DB writes, and
+its result is checked for the `worker_conflict` return.
+
+# Implementation Details
+
+## Data Model
+
+`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`
+`("couch_jobs", "pending", Type, Priority, JobId) = ""`
```

Comment:
Doesn't the pending queue job item need a Jobops section in the value space? How else would we define what the job is about?

Reply:
When the JobId is added to the pending queue it won't be removed from the main jobs table.